design/toolchain.md

```diff
@@ -0,0 +1,115 @@
+
+## Requirements
+
+* It should be possible to execute each component in the toolchain individually by feeding it the output from the previous step.
+* Given a specific input, the output from each step must be completely reproducible (no side effects).
+* Different components may require different execution environments and/or permissions to run. For example `cdk-synth` may need to be able to query the target AWS account in order to resolve environmental context, `cdk-bundle` may need to build docker images, `cdk-deploy` will need admin permissions in order to deploy the app.
+* It should be possible to invoke each component as a jsii library from all supported languages.
+
+## Design
+
+Let's start by formally defining the states and steps in a CDK application's lifecycle:
+
+```
++-----+                +---------+             +----------+             +------------+
+| App |-- synthesis -->| Staging |-- bundle -->| Assembly |-- deploy -->| Deployment |
++-----+                +---------+             +----------+             +------------+
+```
+
+### cdk-synth
+
+We start from a ready-to-run (__compiled__) CDK app.
+
+Some languages require compilation (such as Java, Go, TypeScript) and in others (JavaScript, Ruby, Python), there is no need to compile. Compiling code is __not in scope__ for the CDK toolchain in order to allow developers to use their favorite IDEs and build toolchain idiomatically.
+
+The first component in the toolchain is called `cdk-synth`. It's purpose is to execute the CDK app so it can synthesize the CDK tree (through the internal app phases which are "init", "prepare", "validate", "synthesize"). Eventually the CDK tree can synthesize both final artifacts such as CloudFormation templates and intermediate artifacts such as unpackaged assets.
+
+The protocol between the CDK app `cdk-synth`:
+
+- Context is passed into the app via a JSON map serialized into an environment variable called `CONTEXT_ENV`.
+- The `CDK_OUTDIR` points to a directory into which artifacts should be emitted by the app.
+
+The app will emit into `CDK_OUTDIR` a directory structure as follows:
+
+```
+.
+├── missing.json
+├── bundle.json
+├── assembly
+│   ├── manifest.json
+│   └── stack1.template.json
+└── staging
+    └── intermediate-asset
+        └── file1.txt
+```
+
+If the file `missing.json` exists in the output, it means that the synthesis output cannot be used and relies on missing context information, which must be populated in the context map. `cdk-synth` is responsible to resolve this context information through the environmental context provider system (for example, query the account for which availability zones are available in a specific region, read SSM parameters, etc). Then, `cdk-synth` must re-invoke the app with this context. Bear in mind that normally context information is cached in `cdk.context.json`, so subsequent executions will not require resolution.
```

Comment:
Can `missing.json` be empty? I'd rather use constant work (always generate the standard files) instead of having to conditionally do stuff... *Missing nothing* is something that could be explicitly stated, and this also makes it nicer to catch tools that created a file with a typo in the name or in the wrong location (instead of assuming everything was okay).

Reply:
I actually prefer the opposite - if `missing.json` or `bundle.json` isn't there, it means that the entire phase can be skipped. In the `missing` case it's less evident, but with `bundle.json` it literally means that we can totally skip the bundling phase because we already have a fully-realized assembly.